url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/compiler/parser/XTokenizer.java

```diff
@@ -0,0 +1,385 @@
+			acPos++;
+		}
+		return token;
+	}
+	
+	private XToken tryGetNext2(){
+		spaces = 0;
+		newLines = 0;
+		while(true){
+			while(c==' '||c=='\t'||c=='\r'||c=='\n'){
+				if(c==' '||c=='\t'){
+					spaces++;
+				}else if(c=='\n'){
+					newLines++;
+				}
+				nextChar();
+			}
+			start = fileReader.getPosition();
+			if(c==EOF){
+				return makeToken(XTokenKind.EOF, null);
+			}else if(c=='"' || c=='\''){
+				return makeString(c=='\'');
+			}else if(charOkForRadix(10, c)){
+				return makeNumber(false);
+			}else if(Character.isAlphabetic((char)c) || c=='_'){
+				return makeIdent();
+			}else if(c=='/'){
+				nextChar();
+				if(c=='/' || c=='*'){
+					readLineComment(c=='*');
+				}else{
+					return makeToken(XTokenKind.KEYWORD, XKeyword.DIV);
+				}
+			}else{
+				if(c=='.'){
+					nextChar();
+					if(charOkForRadix(10, c)){
+						return makeNumber(true);
+					}
+					return makeToken(XTokenKind.KEYWORD, XKeyword.DOT);
+				}else{
+					XKeyword kw = XKeyword.getKeyword(Character.toString((char)c));
+					if(kw!=null){
+						nextChar();
+						return makeToken(XTokenKind.KEYWORD, kw);
+					}else{
+						addDiagnostic2("char.unknown", (char)c);
+						nextChar();
+					}
+				}
+			}
+		}
+	}
+	
+	private XToken makeString(boolean makeChar){
+		nextChar();
+		String string = "";
+		char e = makeChar?'\'':'"';
+		while(c!=e && c!=EOF){
+			if(c=='\\'){
+				nextChar();
+				if(c=='b'){
+					string += '\b';
+				}else if(c=='t'){
+					string += '\t';
+				}else if(c=='n'){
+					string += '\n';
+				}else if(c=='r'){
+					string += '\r';
+				}else if(c=='f'){
+					string += '\f';
+				}else if(c=='"'){
+					string += '\"';
+				}else if(c=='\''){
+					string += '\'';
+				}else if(c=='\\'){
+					string += '\\';
+				}else{
+					addDiagnostic((makeChar?"char":"string")+".invalid.escape", (char)c);
+				}
+			}else{
+				string += (char)c;
+			}
+			nextChar();
+		}
+		if(c==EOF){
+			addDiagnostic((makeChar?"char":"string")+".unexpected.eof");
+		}else{
+			nextChar();
+			if(makeChar && string.length()!=1){
+				if(string.length()==0){
+					addDiagnostic("char.invalid.empty");
+				}else{
+					addDiagnostic("char.invalid.length");
+				}
+			}
+		}
+		if(makeChar){
+			long dc = 0;
+			if(string.length()>0){
+				dc = string.charAt(0);
+			}
+			return makeToken(XTokenKind.INT, dc);
+		}
+		return makeToken(XTokenKind.STRING, string);
+	}
+	
+	private static boolean charOkForRadix(int radix, int c){
+		if(radix==2){
+			return c=='0' || c=='1';
+		}else if(radix==8){
+			return c>='0' && c<='7';
+		}else if(radix==10){
+			return c>='0' && c<='9';
+		}else if(radix==16){
+			return (c>='a' && c<='f') || (c>='A' && c<='F') || (c>='0' && c<='9');
+		}
+		return false;
+	}
+	
+	private String scannDigit(int radix){
+		String number = "";
+		while(charOkForRadix(radix, c) || c=='_'){
+			if(c!='_'){
+				number += (char)c;
+			}
+			nextChar();
+		}
+		return number;
+	}
+	
+	private XToken makeNumber(boolean dot){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 29. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 246-275 into a separate method.